server/src-lib/Hasura/GraphQL/Execute.hs

```diff
@@ -747,44 +692,40 @@ arrayIndexText (ArrayIndex i) =
 
 -- | Produce a field from the nested field calls.
 fieldCallsToField ::
-     Maybe G.Alias
-  -> Map.HashMap G.Name AnnInpVal
+     Map.HashMap G.Name AnnInpVal
   -> Map.HashMap G.Variable G.ValueConst
   -> SelSet
+  -> Maybe G.Alias
   -> NonEmpty FieldCall
   -> Field
-fieldCallsToField mindexedAlias0 userProvidedArguments variables finalSelSet =
-  nest mindexedAlias0
+fieldCallsToField userProvidedArguments variables finalSelSet = nest
   where
-    nest mindexedAlias (fieldCall :| rest) =
+    nest mindexedAlias (FieldCall{..} :| rest) =
       Field
-        { _fAlias =
-            case mindexedAlias of
-              Just indexedAlias -> indexedAlias
-              Nothing           -> G.Alias (fcName fieldCall)
-        , _fName = fcName fieldCall
+        { _fAlias = fromMaybe (G.Alias fcName) mindexedAlias
+        , _fName = fcName
         , _fType = G.NamedType (G.Name "unknown_type")
         , _fArguments =
             let templatedArguments =
-                  createArguments variables (fcArguments fieldCall)
+                  createArguments variables fcArguments
+             -- TODO this looks sketchy to me... should we really be checking some invariant here,
+             -- that relates userProvidedArguments and the 'NonEmpty FieldCall'?
              in case NE.nonEmpty rest of
                   Just {} -> templatedArguments
                   Nothing ->
                     Map.unionWith
                       mergeAnnInpVal
                       userProvidedArguments
                       templatedArguments
-        , _fSelSet =
-            case NE.nonEmpty rest of
-              Nothing    -> finalSelSet
-              Just calls -> pure (nest Nothing calls)
+        , _fSelSet = maybe finalSelSet (pure . nest Nothing) $ NE.nonEmpty rest
         , _fRemoteRel = Nothing
         }
 
 mergeAnnInpVal :: AnnInpVal -> AnnInpVal -> AnnInpVal
 mergeAnnInpVal an1 an2 =
   an1 {_aivValue = mergeAnnGValue (_aivValue an1) (_aivValue an2)}
 
+-- TODO what is an AnnGValue? What does the "Ann" prefix signify? Why is it okay to merge them in this particular way?
```

Comment:
good question. This part I haven't looked at myself. Must take Rakesh's help here.

Reply:
Actually now I do remember what this is. This is a kind of "deep merge". For e.g. suppose the input argument of the remote field is something like:

Part of join configuration: `where: { id : 1}`

And during execution, client also gives the input arg: `where: {name: "tiru"}`. So the type of input argument is same, say, `WhereInpArg`, then if we do a naive merge only `where: {id : 1}` will be included whereas we need this: `where: {id : 1, name: "tiru"}`

Reply:
Okay, hm. Maybe when I understand this a little better I can propose a better name, but something referring to "conjunction" or "and", and "predicates" seems clearer